be/src/column/binary_column.cpp

```diff
@@ -373,38 +373,37 @@ void BinaryColumn::deserialize_and_append_batch(std::vector<Slice>& srcs, size_t
     }
 }
 
-uint8_t* BinaryColumn::serialize_column(uint8_t* dst) {
+bool BinaryColumn::serialize_column(io::ZeroCopyOutputStream* out) {
+    io::CodedOutputStream os(out);
     uint32_t bytes_size = _bytes.size() * sizeof(uint8_t);
-    encode_fixed32_le(dst, bytes_size);
-    dst += sizeof(uint32_t);
-
-    strings::memcpy_inlined(dst, _bytes.data(), bytes_size);
-    dst += bytes_size;
-
     uint32_t offsets_size = _offsets.size() * sizeof(Offset);
-    encode_fixed32_le(dst, offsets_size);
-    dst += sizeof(uint32_t);
-
-    strings::memcpy_inlined(dst, _offsets.data(), offsets_size);
-    dst += offsets_size;
-    return dst;
+    int buff_size = static_cast<int>(sizeof(bytes_size) + sizeof(offsets_size) + bytes_size + offsets_size);
+    uint8_t* buff = os.GetDirectBufferForNBytesAndAdvance(buff_size);
+    if (buff != nullptr) {
+        buff = io::CodedOutputStream::WriteLittleEndian32ToArray(bytes_size, buff);
+        buff = io::CodedOutputStream::WriteRawToArray(_bytes.data(), static_cast<int>(bytes_size), buff);
+        buff = io::CodedOutputStream::WriteLittleEndian32ToArray(offsets_size, buff);
+        buff = io::CodedOutputStream::WriteRawToArray(_offsets.data(), static_cast<int>(offsets_size), buff);
```

Comment:
offsets_size could be uint_32 max

Reply:
> 
Since each element in `_offsets` column is 4 bytes, a value of `offsets_size` greater than  INT_MAX means that the `_offsets` column contains more than 536870911 elements, which is impossible in practice I think.

Reply:
> > 
> 
> Since each element in `_offsets` column is 4 bytes, a value of `offsets_size` greater than INT_MAX means that the `_offsets` column contains more than 536870911 elements, which is impossible in practice I think.

If you think this is still a practical problem I can fix this with additional checks.

Reply:
You are right.  I wanted to point to bytes_size.
```
    bool reach_capacity_limit() const override {
        return _bytes.size() >= Column::MAX_CAPACITY_LIMIT || _offsets.size() >= Column::MAX_CAPACITY_LIMIT ||
               _slices.size() >= Column::MAX_CAPACITY_LIMIT;
    }
```
In join, build hash table. _bytes size could be uint_32 max.

Reply:
> You are right. I wanted to point to bytes_size.
> 
> ```
>     bool reach_capacity_limit() const override {
>         return _bytes.size() >= Column::MAX_CAPACITY_LIMIT || _offsets.size() >= Column::MAX_CAPACITY_LIMIT ||
>                _slices.size() >= Column::MAX_CAPACITY_LIMIT;
>     }
> ```
> 
> In join, build hash table. _bytes size could be uint_32 max.

Got it, I'll fix it